Roles support

bkimbrough88 opened the pull request:
Enabling the SDK to perform CRUD operations on Roles as well as manage permissions and users on those roles.

Fixes: #307

Commit:
ISSUE-307 - Initial commit to support roles

* Added code to allow for performing CRUD operations on a role
* Added code to allow granting and revoking permissions on a role
* Added code to add or remove users on a role
* Initial unit tests

Commit:
ISSUE-307 - adding more unit tests

* Added remaining role CRUD tests
* Added permissions tests

Commit:
ISSUE-307 - Finishing vanilla unit tests

* Ensuring all role methods have at least one unit test
* Splitting permissions and organizations data types and methods into their own files

Commit:
ISSUE-307 - Removing idea files

sendqueery commented:
This looks good, but appears to be missing support for scoping unless I've missed something?
It looks like it's missing from the API docs, but is mentioned elsewhere in Datadog's docs: https://docs.datadoghq.com/account_management/faq/managing-global-role-permissions/?tab=datadogussite#granting-permissions-within-limited-scopes

From their example:

```
{
    "data": {
        "type": "permissions",
         "id": <PERMISSION_UUID>,
         "scope": {
             "indexes": [
                 "main",
                 "support"
            0]
        }
    }
}
```

Commit:
ISSUE-307

Commit:
ISSUE-307 - adding scope to permission grant requests

bkimbrough88 commented:
You're right, I've added permission scope to the grant permission requests

Commit:
ISSUE-307 - making scope be passed in as a pointer

pietdaniel commented:
Some of the APIs implemented are not completely stable yet. My concern is that we may be solidifying prematurely here. Namely, the scope field and the org v1 APIs. Is it possible to postpone the implementation of this functionality? the role and permission APIs can be considered stable.

Commit:
ISSUE-307 - Removing .dependencies and .env files

These files should not have ever been added
Also enhancing git ignore file to ensure other useless files don't get added

bkimbrough88 commented on .dependencies/src/github.com/cenkalti/backoff:
```diff
@@ -0,0 +1 @@
+Subproject commit 4b4cebaf850ec58f1bb1fec5bdebdf8501c2bc3f
```
yeah didn't notice that, removed it, not sure when or why that one got added

jirikuncar replied:
Thanks!

jirikuncar commented on .gitignore:
```diff
@@ -1,2 +1,239 @@
-*.sublime*
-cmd
+# Created by https://www.gitignore.io/api/go,vim,linux,macos,windows,sublimetext,intellij+iml,visualstudiocode
```
Do you think we could create a separate PR just for a `.gitignore`?

jirikuncar commented on roles_test.go:
```diff
@@ -0,0 +1,665 @@
+			t.Fatal(err)
+		} else {
+			var roleRequest map[string]*Role
+			println(string(requestBody))
```
do we still need it?

bkimbrough88 replied:
good eye, that should not be there

Commit:
ISSUE-307  - Addressing PR feedback

* Removed unnecessary print statement
* Reverted git ignore to previous state, will open a new PR with those changes

jirikuncar commented:
@bkimbrough88 is there a way how we could reduce the diff in `datadog-accessors.go` file (maybe reordering methods) to facilitate the review process?

jirikuncar commented on organizations.go:
```diff
@@ -0,0 +1,11 @@
+	Data *Organization `json:"data,omitempty"`
+}
+
+// TODO: Fill this out with everything you can get back when getting an organization
```
❓

bkimbrough88 replied:
This was just a comment that there's much more to Organizations than an ID, but that's all I needed for this implementation. I can easily delete the comment or create an issue for adding Organizations support.

pietdaniel reviewed (changes requested):
I worry that it is a bit early for the role permission API to be solidified in the public client. Can we postpone implementing this API while we roadmap the APIs definition internally. Roles Users and Permissions are good to go.

jirikuncar commented on roles.go:
```diff
@@ -0,0 +1,198 @@
+type Sort string
+
+const (
+	SortNameAsc        Sort = "name"
+	SortNameDesc       Sort = "-name"
+	SortModifiedAtAsc  Sort = "modified_at"
+	SortModifiedAtDesc Sort = "-modified_at"
+	SortUserCountAsc   Sort = "user_count"
+	SortUserCountDesc  Sort = "-user_count"
```
What about adding `SortRolesBy...`?

bkimbrough88 commented:
> I worry that it is a bit early for the role permission API to be solidified in the public client. Can we postpone implementing this API while we roadmap the APIs definition internally. Roles Users and Permissions are good to go.

I feel like anything posted in https://docs.datadoghq.com/api should be stable enough to write into an SDK. The caveat being that you then have to maintain that feature in the SDK as the API evolves.

bkimbrough88 commented:
Apologies on letting this sit, I'll take care of the PR comments later tonight

Commit:
removing scoped roles and other PR feedback
